### DISCUSSION TITLE
a potential performance improvement in message parser

### DISCUSSION THREAD

**Discussion by zmstone** [vernemq/vernemq#1812] [Ideas]
Hi There.

There is potentially a performance killer in this line:
https://github.com/vernemq/vernemq/blob/ddeeda4864aaf26381fc620cb5b98c93bed5d83e/apps/vmq_server/src/vmq_ranch.erl#L201
When the packet size is very large, this binary concatenation becomes more and more expensive especially when the tcp data comes in in small pieces. 

Sorry I might not have the full picture of the code, and I'm just too lazy to verify it.

**Comment by ioolkos** [comment]
@zmstone thanks for the idea :) will try to check & measure.
